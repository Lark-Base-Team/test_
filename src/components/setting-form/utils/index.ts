import { FieldType } from "@lark-base-open/js-sdk";

export const ignoreFieldType = [
  FieldType.Lookup,
  FieldType.CreatedTime,
  FieldType.ModifiedTime,
  FieldType.Formula,
  FieldType.DuplexLink,
  FieldType.SingleLink,
  FieldType.CreatedUser,
  FieldType.ModifiedUser,
  FieldType.NotSupport,
  FieldType.Location,
  FieldType.AutoNumber,
  FieldType.GroupChat,
  // FieldType.User,
  FieldType.Denied,
  FieldType.Attachment,
];
