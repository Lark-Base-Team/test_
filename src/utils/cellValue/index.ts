import {
  IOpenCellValue,
  FieldType,
  IMultiSelectFieldMeta,
  ISingleSelectFieldMeta,
} from "@lark-base-open/js-sdk";
import { fieldMap } from "@/types/types";
import { dateTime, dateDefaultFormat } from "./date";
import { multiSelect } from "./multiSelect";
import { singleSelect } from "./singleSelect";
import { checkBox, defaultBoolValue } from "./checkBox";
import { text } from "./text";
import { url } from "./url";
import { phone } from "./phone";
import { currency } from "./currency";
import { progress } from "./progress";
import { rating } from "./rating";
import { barCode } from "./barCode";
import {User} from "./user";

export function getCellValue(
  fieldMap: fieldMap,
  value: string,
): IOpenCellValue {
  const type = fieldMap.field.type;
  const field = fieldMap.field;
  const config = fieldMap.config;
  if (!value) return value;
  switch (type) {
    case FieldType.Url:
      return url(value);
    case FieldType.DateTime:
      return dateTime(value, config?.format ?? dateDefaultFormat);
    case FieldType.SingleSelect:
      return singleSelect(value, field as ISingleSelectFieldMeta);
    case FieldType.MultiSelect:
      return multiSelect(
        value,
        field as IMultiSelectFieldMeta,
        config?.separator ?? ","
      );
    case FieldType.Number:
      return Number(value.match(/-?\d+\.?\d*/g));
    case FieldType.Currency:
      return currency(value);
    case FieldType.Progress:
      return progress(value);
    case FieldType.Rating:
      return rating(value);
    case FieldType.Checkbox:
      return checkBox(value, config?.boolValue ?? defaultBoolValue);
    case FieldType.Text:
      return text(value) as IOpenCellValue;
    case FieldType.Barcode:
      return barCode(value) as IOpenCellValue;
    case FieldType.Phone:
      return phone(value);
    case FieldType.User:
      return User(value);
    default:
      return value;
  }
}
