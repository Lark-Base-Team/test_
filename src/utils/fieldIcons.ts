import { Calendar, Phone, User } from "@element-plus/icons-vue"
import { FieldType } from "@lark-base-open/js-sdk"
import textIcon from "@/components/icons/text-icon.vue"
import barCodeIcon from "@/components/icons/barCode-icon.vue"
import multiSelectIcon from "@/components/icons/multiSelect-icon.vue"
import checkBoxIcon from "@/components/icons/checkBox-icon.vue"
import singleSelectIcon from "@/components/icons/singleSelect-icon.vue"
import currencyIcon from "@/components/icons/currency-icon.vue"
import numberIcon from "@/components/icons/number-icon.vue"
import progressIcon from "@/components/icons/progress-icon.vue"
import ratingIcon from "@/components/icons/rating-icon.vue"
import urlIcon from "@/components/icons/url-icon.vue"
import singleLinkIcon from "@/components/icons/singleLink-icon.vue"
import duplexLinkIcon from "@/components/icons/duplexLink-icon.vue"

export const iconList: { [key: number]: any } = {
  [FieldType.DateTime]: Calendar,
  [FieldType.Checkbox]: checkBoxIcon,
  [FieldType.Phone]: Phone,
  [FieldType.Text]: textIcon,
  [FieldType.Barcode]: barCodeIcon,
  [FieldType.MultiSelect]: multiSelectIcon,
  [FieldType.SingleSelect]: singleSelectIcon,
  [FieldType.Currency]: currencyIcon,
  [FieldType.Number]: numberIcon,
  [FieldType.Progress]: progressIcon,
  [FieldType.Rating]: ratingIcon,
  [FieldType.Url]: urlIcon,
  [FieldType.User]: User,
  [FieldType.SingleLink]: singleLinkIcon,
  [FieldType.DuplexLink]: duplexLinkIcon,
}
