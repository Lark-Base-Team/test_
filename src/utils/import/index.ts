import { importModes, importExcel } from "./import";
// import { importWorkerInstance } from "./worker";
import {
  importLifeCircles,
  addLifeCircleEvent,
  lifeCircleEvent,
  removeLifeCircleEvent,
  clearLifeCircleEvent,
  runLifeCircleEvent,
  lifeCircleEventParams
} from "./lifeCircle";

export {
  importModes,
  // importWorkerInstance,
  importLifeCircles,
  addLifeCircleEvent,
  removeLifeCircleEvent,
  clearLifeCircleEvent,
  runLifeCircleEvent,
  importExcel,
  
};
export type { lifeCircleEvent, lifeCircleEventParams };
