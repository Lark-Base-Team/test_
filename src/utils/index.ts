import { useDark } from "@vueuse/core"
export * from "./useBitable"

export const isDark = useDark()
