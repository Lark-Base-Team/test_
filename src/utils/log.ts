import { ref } from "vue"

export type LogOptions = {
  prefix: string
  onError?: Function
  onWarn?: Function
  onInfo?: Function
  onLog?: Function
}

export type LogItem = {
  type: "info" | "warn" | "error" | "log"
  message: string
  time: number
  file?: string
  line?: number
}

export const log = ref<LogItem[]>([])

export const options = ref<LogOptions>({
  prefix: "[Excel Compare and Import]",
})

const _console = console

export const useLog = (type: LogItem["type"]) => (msg: string) => {
  const cache = {
    type: type,
    message: msg,
    time: Date.now(),
  }
  log.value.push(cache)
  const action = _console[type] ?? _console.log
  action.apply(void 0, [options.value.prefix, msg])
  if (type === "error") {
    options.value.onError?.(msg)
  }

  if (type === "warn") {
    options.value.onWarn?.(msg)
  }

  if (type === "info") {
    options.value.onInfo?.(msg)
  }

  if (type === "log") {
    options.value.onLog?.(msg)
  }
}
