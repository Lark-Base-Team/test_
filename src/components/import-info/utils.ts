import { ref } from "vue";
import { i18n } from "@/i18n";
import { lifeCircleEventParams } from "@/utils/import/lifeCircle";

export interface stage {
  state: string;
  disabled: boolean;
  message?: string;
  progress: boolean;
  stage: string;
  success: number;
  error: number;
  number: number;
  name: string | number;
  title: string;
  index: number;
}

const stageIndex = {
  checkFieldTypes: 0,
  checkOptions: 1,
  setOptions: 2,
  analysisRecords: 3,
  updateRecords: 4,
  deleteRecords: 5,
  addRecords: 6,
};

export const currentStage = ref(0);

export const defaultStages = () => [
  {
    state: "waiting",
    disabled: false,
    message: "",
    progress: true,
    stage: "checkFieldTypes",
    success: 0,
    error: 0,
    number: 0,
    name: stageIndex["checkFieldTypes"],
    title: "importInfo.checkFieldTypes",
    index: stageIndex["checkFieldTypes"],
  },
  {
    state: "waiting",
    disabled: true,
    message: "",
    progress: false,
    stage: "checkOptions",
    success: 0,
    error: 0,
    number: 0,
    name: stageIndex["checkOptions"],
    title: "importInfo.checkOptions",
    index: stageIndex["checkOptions"],
  },
  // {
  //   state: "waiting",
  //   disabled: true,
  //   message: "",
  //   progress: false,
  //   stage: "setOptions",
  //   success: 0,
  //   error: 0,
  //   number: 0,
  //   name: stageIndex["setOptions"],
  //   title: "importInfo.setOptions",
  //   index: stageIndex["setOptions"],
  // },
  {
    state: "waiting",
    disabled: true,
    message: "",
    progress: false,
    stage: "analysisRecords",
    success: 0,
    error: 0,
    number: 0,
    name: stageIndex["analysisRecords"],
    title: "importInfo.analysisRecords",
    index: stageIndex["analysisRecords"],
  },
  {
    state: "waiting",
    disabled: true,
    message: "",
    progress: false,
    stage: "updateRecords",
    success: 0,
    error: 0,
    number: 0,
    name: stageIndex["updateRecords"],
    title: "importInfo.updateRecords",
    index: stageIndex["updateRecords"],
  },
  {
    state: "waiting",
    disabled: true,
    message: "",
    progress: false,
    stage: "deleteRecords",
    success: 0,
    error: 0,
    number: 0,
    name: stageIndex["deleteRecords"],
    title: "importInfo.deleteRecords",
    index: stageIndex["deleteRecords"],
  },
  {
    state: "waiting",
    disabled: true,
    message: "",
    progress: false,
    stage: "addRecords",
    success: 0,
    error: 0,
    number: 0,
    name: stageIndex["addRecords"],
    title: "importInfo.addRecords",
    index: stageIndex["addRecords"],
  },
];

export const stages = ref<stage[]>(defaultStages());

export function beforeCheckFields(e: lifeCircleEventParams) {
  const { number } = e.data;
  const index = stages.value.findIndex(
    (stage) => stage.index === stageIndex["checkFieldTypes"]
  );
  stages.value[index].number = number;
  if (number > 0) {
    stages.value[index].state = "loading";
    stages.value[index].progress = true;
  } else {
    stages.value[index].state = "success";
    stages.value[index].progress = false;
  }

  stages.value[index].disabled = false;
  currentStage.value = stageIndex["checkFieldTypes"];
}

const optionsFieldsNumber = ref(0);

export function onCheckFields(e: lifeCircleEventParams) {
  const { res } = e.data;
  const index = stages.value.findIndex(
    (stage) => stage.index === stageIndex["checkFieldTypes"]
  );
  stages.value[index].success += 1;
  if (res) {
    optionsFieldsNumber.value += 1;
    stages.value[index].message = i18n.global.t(
      "importInfo.checkFieldsMessage",
      {
        totalNumber: optionsFieldsNumber.value,
      }
    );
  }
  if (
    stages.value[index].success + stages.value[index].error ===
    stages.value[index].number
  ) {
    stages.value[index].progress = false;
    stages.value[index].state = "success";
  }
}

export function beforeCheckOptions(e: lifeCircleEventParams) {
  const { number } = e.data;
  const index = stages.value.findIndex(
    (stage) => stage.index === stageIndex["checkOptions"]
  );
  stages.value[index].number = number;
  if (number > 0) {
    stages.value[index].state = "loading";
    stages.value[index].progress = true;
  } else {
    stages.value[index].state = "success";
    stages.value[index].progress = false;
  }
  stages.value[index].disabled = false;
  currentStage.value = stageIndex["checkOptions"];
}

export function onCheckOptions(e: lifeCircleEventParams) {
  const { selects } = e.data;
  const index = stages.value.findIndex(
    (stage) => stage.index === stageIndex["checkOptions"]
  );
  console.log("onCheckOptions", selects, stages.value[index]);
  stages.value[index].success += 1;
  stages.value[index].message = i18n.global.t(
    "importInfo.checkOptionsMessage",
    {
      totalNumber: selects.length,
    }
  );
  if (
    stages.value[index].success + stages.value[index].error ===
    stages.value[index].number
  ) {
    stages.value[index].progress = false;
    stages.value[index].state = "success";
  }
}

// export function beforeSetOptions(e: lifeCircleEventParams) {
//   const { number } = e.data;
//   const index = stages.value.findIndex(
//     (stage) => stage.index === stageIndex["setOptions"]
//   );
//   stages.value[index].number = number;
//   if (number > 0) {
//     stages.value[index].state = "loading";
//     stages.value[index].progress = true;
//   } else {
//     stages.value[index].state = "success";
//   }
//   stages.value[index].disabled = false;
//   currentStage.value = stageIndex["setOptions"];
// }

// export function onSetOptions(e: lifeCircleEventParams) {
//   const { record } = e.data;
//   const index = stages.value.findIndex(
//     (stage) => stage.index === stageIndex["setOptions"]
//   );
//   if (record) {
//     stages.value[index].success += 1;
//   } else {
//     stages.value[index].error += 1;
//   }
// }

export function beforeAnalysisRecords(e: lifeCircleEventParams) {
  const { number } = e.data;
  const index = stages.value.findIndex(
    (stage) => stage.index === stageIndex["analysisRecords"]
  );
  stages.value[index].number = number;
  if (number > 0) {
    stages.value[index].state = "loading";
    stages.value[index].progress = true;
  } else {
    stages.value[index].state = "success";
    stages.value[index].progress = false;
  }
  stages.value[index].disabled = false;
  currentStage.value = stageIndex["analysisRecords"];
}

export function onAnalysisRecords(e: lifeCircleEventParams) {
  const { updateNumber, addNumber, message } = e.data;
  const index = stages.value.findIndex(
    (stage) => stage.index === stageIndex["analysisRecords"]
  );
  stages.value[index].success = updateNumber + addNumber;
  stages.value[index].message = message;
  if (stages.value[index].success === stages.value[index].number) {
    stages.value[index].progress = false;
    stages.value[index].state = "success";
  }
}

export function beforeUpdateRecords(e: lifeCircleEventParams) {
  const { updateList } = e.data;
  const index = stages.value.findIndex(
    (stage) => stage.index === stageIndex["updateRecords"]
  );
  stages.value[index].number = updateList.length;
  if (updateList.length > 0) {
    stages.value[index].state = "loading";
    stages.value[index].progress = true;
  } else {
    stages.value[index].state = "success";
    stages.value[index].progress = false;
  }
  stages.value[index].disabled = false;
  currentStage.value = stageIndex["updateRecords"];
}

export function onUpdateRecords(e: lifeCircleEventParams) {
  const { res } = e.data;
  const index = stages.value.findIndex(
    (stage) => stage.index === stageIndex["updateRecords"]
  );
  if (res) {
    stages.value[index].success += 500;
  } else {
    stages.value[index].error += 500;
  }
  if (
    stages.value[index].success + stages.value[index].error ===
    stages.value[index].number
  ) {
    stages.value[index].progress = false;
    stages.value[index].state = "success";
  }
}

export function beforeDeleteRecords(e: lifeCircleEventParams) {
  const { deleteList } = e.data;
  const index = stages.value.findIndex(
    (stage) => stage.index === stageIndex["deleteRecords"]
  );
  stages.value[index].number = deleteList.length;
  if (deleteList.length > 0) {
    stages.value[index].state = "loading";
    stages.value[index].progress = true;
  } else {
    stages.value[index].state = "success";
    stages.value[index].progress = false;
  }
  stages.value[index].disabled = false;
  currentStage.value = stageIndex["deleteRecords"];
}

export function onDeleteRecords(e: lifeCircleEventParams) {
  const { res } = e.data;
  const index = stages.value.findIndex(
    (stage) => stage.index === stageIndex["deleteRecords"]
  );
  if (res) {
    stages.value[index].success += 500;
  } else {
    stages.value[index].error += 500;
  }
  if (
    stages.value[index].success + stages.value[index].error ===
    stages.value[index].number
  ) {
    stages.value[index].progress = false;
    stages.value[index].state = "success";
  }
}

export function beforeAddRecords(e: lifeCircleEventParams) {
  const { number } = e.data;
  const index = stages.value.findIndex(
    (stage) => stage.index === stageIndex["addRecords"]
  );
  stages.value[index].number = number;
  stages.value[index].message = i18n.global.t("importInfo.addRecordsMessage");
  if (number > 0) {
    stages.value[index].state = "loading";
    stages.value[index].progress = true;
  } else {
    stages.value[index].state = "success";
    stages.value[index].progress = false;
  }
  stages.value[index].disabled = false;
  currentStage.value = stageIndex["addRecords"];
}

export function onAddRecords(e: lifeCircleEventParams) {
  const { res } = e.data;
  const index = stages.value.findIndex(
    (stage) => stage.index === stageIndex["addRecords"]
  );
  if (res) {
    stages.value[index].success += 500;
  } else {
    stages.value[index].error += 500;
  }
  if (
    stages.value[index].success + stages.value[index].error ===
    stages.value[index].number
  ) {
    stages.value[index].state = "success";
  }
}
