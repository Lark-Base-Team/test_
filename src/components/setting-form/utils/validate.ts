import type { fieldMap } from "@/types/types"
import { Warn } from "@/utils"

export function validateIndex(index: fieldMap[]) {
  if (index.length === 0) {
    Warn({
      title: "noIndex",
      message: "noIndex",
      notice: true,
      noticeParams: {
        text: "message.noIndex",
      },
    })
    return false
  }
  const noExcelField = index.filter((i) => !i.excel_field)
  if (noExcelField.length) {
    const names = noExcelField.map((i) => i.field.name).join(", ")
    Warn({
      title: "noExcelField",
      message: names,
      notice: true,
      noticeParams: {
        text: "message.indexNoExcelField",
        params: {
          fields: names,
        },
      },
    })
    return false
  }
  return true
}
