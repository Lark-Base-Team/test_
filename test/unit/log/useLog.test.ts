import { useLog, log } from "../../../src/utils/log"
import { describe, it, expect, beforeEach, vi, afterEach } from "vitest"

describe("useLog", () => {
  beforeEach(() => {
    vi.useFakeTimers()
  })

  afterEach(() => {
    vi.useRealTimers()
  })

  it("should return a function", () => {
    expect(typeof useLog("info")).toBe("function")
  })

  it("should push a LogItem to log.value", () => {
    const date = new Date(2000, 1, 1, 13)
    vi.setSystemTime(date)
    const logInfo = useLog("info")
    logInfo("Hello world")
    expect(log.value[0]).toMatchSnapshot()
    const logWarn = useLog("warn")
    logWarn("Hello world")
    expect(log.value[1]).toMatchSnapshot()
    const logError = useLog("error")

    logError("Hello world")
    expect(log.value[2]).toMatchSnapshot()
    const logLog = useLog("log")
    logLog("Hello world")
    expect(log.value[3]).toMatchSnapshot()
  })
})
