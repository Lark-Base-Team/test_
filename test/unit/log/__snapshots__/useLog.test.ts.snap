// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useLog > should push a LogItem to log.value 1`] = `
{
  "message": "Hello world",
  "time": 949381200000,
  "type": "info",
}
`;

exports[`useLog > should push a LogItem to log.value 2`] = `
{
  "message": "Hello world",
  "time": 949381200000,
  "type": "warn",
}
`;

exports[`useLog > should push a LogItem to log.value 3`] = `
{
  "message": "Hello world",
  "time": 949381200000,
  "type": "error",
}
`;

exports[`useLog > should push a LogItem to log.value 4`] = `
{
  "message": "Hello world",
  "time": 949381200000,
  "type": "log",
}
`;
